### DISCUSSION TITLE
A node based GUI?

### DISCUSSION THREAD

**Discussion by SMUsamaShah** [zero01101/openOutpaint#162] [Ideas]
I have this idea for quite a while (since I know about this project) and have looked into it a bit too but never got the courage to actually do it, not very good at it anyway. Just putting it out here to see if it is really that useful.

What if this tool was a node based interface or had that as an alternative UI? Would it have made it more convenient for everyone. Only thing I am sure about is that all components would have become tiny modules which you could connect and visually see your pipeline too.

For example you would have nodes to engineer prompts. Then add further prompts on a result. Or expand (outpaint) a result or apply img2img on it. All of these steps could have been visible as previews.

I thought there has to be something like this already. I did find some tools like that. There are desktop applications but some web applications too. https://gist.github.com/SMUsamaShah/c923a0af4543ee2746979328825da236
 2, 5 and 7 from this list are interesting ones. 

I really think there should exist a node based GUI for SD. If there is something like that already I don't about it.

**Comment by seijihariki** [comment]
Hi, do you mean node-based blender style? If so, could you give some basic examples of nodes that would exist, and their inputs/outputs?

**Reply by SMUsamaShah** [reply]
Not used blender so don't know. But there are quite a few tools in that list which you can use right from your browser.

There can be a number of nodes.

Prompt node with a text box
Inputs: Text
Outputs: Text
Settings: append/prepend text

Image generation node
Inputs: Text, CFG, Seed, mask, image, etc
Outputs: Image (shows image thumbnail on node itself)
Settings: img2img, new image
or this could be two different nodes one of img2img other for normal image generation

Outpainting node
Inputs: Image, Seed
Outputs: Image
Settings: directions to expand

Mask node
Inputs: Image
Outputs: Mask
Clicking on it can show a pane to draw the mask over the input image

Clicking each node can show its own interactive panel (with brushes to paint or settings to change). I can not think of a way to mix this with current infinite canvas. There should be way though to have both, that canvas and modularity of nodes.

**Reply by zero01101** [reply]
so like in my own very personal opinion, i don't really see much of a use case for a node editor in here currently; there's only so many "pathways" a workflow can take, and none of the tools involved really have an insurmountable text-based interface expecting three dozen floats or something like a node input would make easier, plus so much of the "well what do i do next" part of SD is so deeply contextual on the output you may or may not have just received...

**Reply by SMUsamaShah** [reply]
Wouldn't nodes with previews make iterations, since that's what we do mostly, lot easier?

As in you create an image, add some prompts and see how that effected original image, or tune CFG or steps and see which direction they lead an image to.

**Reply by seijihariki** [reply]
I... Think I kind of get what you meant with nodes now.

We could try implementing something of the sort as a generation tool, by itself. It would run the node pipeline for each generation. But I don't think it would be very useful if it required any user input during the pipeline itself (masking would be weird, as the pipeline would have to wait for user input. If we had automatic segmentation that would be fine).

Maybe making a [txt2img => upscale => img2img => upscale] workflow could be useful in some ways... But it would be a very niche tool... Adding flow control nodes could be useful to implement something like img2img loop.

I think we need to mature this idea better before implementing it. Defining useful/not useful nodes and sketching some clear workflow ideas.

**Reply by zero01101** [reply]
so ignoring the heresy of mentioning a library in this staunchly, aggressively vanilla javascript only project, https://flume.dev/ is a neat library.  that being said, i STILL really don't think there's enough "composable complexity" in what openOutpaint _does_ that a node editor would provide any benefit, especially if each node has to wait for the previous in the chain to finish (or at least partially finish) processing before anything contextually relevant can be done in the next hop down.  

like check out flume's examples of other node interfaces: they're all things that can be _nearly instantly_ computed, or are a giant pain to manually compose 
![image](https://user-images.githubusercontent.com/1649724/212440851-39a5bb9c-2dd6-422b-a99d-c46920caa936.png) or something like reaktor 
![image](https://user-images.githubusercontent.com/1649724/212440966-fade8ce3-d032-498c-b9d2-8d6e90ce354c.png)


i'm probably just a grumpy old man but it just really feels like it's more cool than useful to how openOutpaint works, especially coming from a (somewhat modular) synth background and learning a proper respect for feedback :| it just feels like it would be better served as its own project?  idunno, i'm not trying to say "absolutely not" or anything, i just really don't see the benefit personally but again, probably just grumpy old man ;)

**Reply by seijihariki** [reply]
I agree that it would probably do better as a separate extension. It simply doesn't seem to fit the current openOutpaint flow.

**Reply by SMUsamaShah** [reply]
I also agree that it can be a separate tool and cannot think of a way to integrate into openOutpaint.
https://nimp.app/ seems like a really nice candidate for that.

**Reply by seijihariki** [reply]
Hey, @SMUsamaShah, don't know if you keep watch of the stable diffusion subreddit, but this:
https://www.reddit.com/gallery/10lzgze
Seems to be nearly exactly what you were wanting, and it's eons ahead of whatever whole could probably pull off using the automatic1111 webui's API.

It seems the UI is this one:
https://github.com/comfyanonymous/ComfyUI

It could be a bit of a pain to switch UIs every so often to use this, but it may be an okay workflow until something comparable is useable in automatic's. Maybe there is an extension somewhere that does this.

**Reply by zero01101** [reply]
very highly seconding comfyUI holy crap that set conditioning area process is BRILLIANT and yeah, this is _precisely_ where a node editor is useful lol; we'd NEVER be able to hook that deeply into the actual diffusion process in this app ;)
